semgrep/semgrep
# refactor: ppx deriving visitors for generic AST

This adds an iter visitor generated by ppx deriving visitors (http://gallium.inria.fr/~fpottier/visitors/manual.pdf) for the generic AST.

To prove it out, I also converted `range_of_any_opt` to use the new visitor. This will be a good proving ground for this approach. Because `range_of_any_opt` is such a hot code path, I was uncertain of the performance implications of using dynamic dispatch here (although to some extent, the current architecture already makes use of explicit dynamic dispatch). Due to this uncertainty, I conducted some performance tests (described below).

I plan to follow up by converting the existing `mk_visitor` function to use the new autogenerated visitors while presenting the same external interface. Then, we can move callsites over one by one and eventually retire that interface.

Test plan:

Existing automated tests

Manual test of performance on juice-shop (rev
c7e1ade8c593875d5fdb6b3b51cef306adf53de9).

Command: `time semgrep-core -l ts -rules default.yml -j 1 .` where `default.yml` is downloaded from https://semgrep.dev/c/p/default.

Methodology: Run command four times both before and after this change, discarding first result for each series of runs. Record "real" time.

Before:

- 56.459s
- 53.678s
- 55.094s

After:

- 56.344s
- 55.342s
- 54.086s

I haven't done a statistical analysis on these results but they look pretty much the same to me.

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)
